lucene/sandbox/src/java/org/apache/lucene/sandbox/rabitq/SpaceUtils.java

```diff
@@ -1,43 +1,60 @@
 package org.apache.lucene.sandbox.rabitq;
 
+import jdk.incubator.vector.ByteVector;
+import jdk.incubator.vector.VectorOperators;
+import jdk.incubator.vector.VectorSpecies;
+import org.apache.lucene.util.BitUtil;
+
 import java.nio.ByteBuffer;
+import java.util.BitSet;
 
 public class SpaceUtils {
 
-  public static int popcount(long[] d, int B) {
-    int ret = 0;
-    for (int i = 0; i < B / 64; i++) {
-      // FIXME: implement this in C because Long.bitCount is not optimized
-      // FIXME: Integer.bitCount used like this is still slower than Long.bitCount
-      // FIXME: a more comprehensive refactor to remove longs and replace with ints is non-trivial
-      ret += Long.bitCount(d[i]);
-      //            ret += Integer.bitCount((int)(d[i] & 0x00000000ffffffffL));
-      //            ret += Integer.bitCount((int)((d[i] & 0xffffffff00000000L) >>> 32));
+  private static final VectorSpecies<Byte> BYTE_SPECIES = ByteVector.SPECIES_PREFERRED;
+
+  private static final int B_QUERY = 4;
+
+  public static int popcount(byte[] d, int B) {
+    return BitSet.valueOf(d).cardinality();
+  }
+
+  public static long ipByteBinByte(byte[] q, byte[] d, int B) {
```

Comment:
lets get rid of `B` all together. We know its always `4` for now.

Reply:
B is a multiple of 64 for D (dimensions) so that one can't go away.